Formatter / schema.org / Add croissant spec 🥐

fxprunayre opened the pull request:
"Croissant 🥐 is a high-level format for machine learning datasets that combines metadata, resource file descriptions, data structure, and default ML semantics into a single file; it works with existing datasets to make them easier to find, use, and support with tools. Croissant builds on schema.org, and its Dataset vocabulary, a widely used format to represent datasets on the Web, and make them searchable." https://docs.mlcommons.org/croissant/

Croissant is extending schema.org, this improvement review the current schema.org formatter to support additional 🥐 metadata available in ISO format. This is mainly about adding:
* croissant fileObject based on online resources with a download protocol
* croissant recordSet based on the feature catalogue

Refactor JSON-LD formatter for using same base formatter for both ISO19139 and ISO19115-3 to facilitate maintenance (similar to citation and DCAT formatter).

Improve formatter producing JSON output by ensuring the output is JSON valid, format it and log any error in order to be able to track errors and improve not well managed encoding.

schema.org improvement:
* `inLanguage` correspond to the resource language, not the metadata language.
* dispatch parties by role instead of only using `producer` (eg. provider, producer, copyrightHolder, publisher, author, funder)
* do not generate element (eg. `temporalCoverage`) if no corresponding element in input document
* add dublin core schema plugin support


Similar initiatives:
* Dataverse https://github.com/gdcc/exporter-croissant/blob/main/src/main/java/io/gdcc/spi/export/croissant/CroissantExporter.java
* CKAN https://github.com/ckan/ckanext-dcat/blob/master/ckanext/dcat/profiles/croissant.py


<!--Include a few sentences describing the overall goals for this Pull Request-->
  
<!-- Please help our volunteers reviewing this PR by completing the following items. 
Ask in a comment if you have troubles with any of them. -->

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geonetwork/core-geonetwork/blob/main/CONTRIBUTING.md)
- [ ] *Pull request* provided for `main` branch, backports managed with label
- [ ] *Good housekeeping* of code, cleaning up comments, tests, and documentation
- [ ] *Clean commit history* broken into understandable chucks, avoiding big commits with hundreds of files, cautious of reformatting and whitespace changes
- [ ] *Clean commit message*s, longer verbose messages are encouraged
- [ ] *API Changes* are identified in commit messages
- [ ] *Testing* provided for features or enhancements using [automatic tests](https://github.com/geonetwork/core-geonetwork/blob/main/software_development/TESTING.md)
- [ ] *User documentation* provided for new features or enhancements in [manual](https://github.com/geonetwork/core-geonetwork/tree/main/docs/manual)
- [ ] *Build documentation* provided for development instructions in `README.md` files
- [ ] *Library management* using `pom.xml` dependency management. Update build documentation with intended library use and library tutorials or documentation

<!--Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or not applicable.-->

<!-- If you can, it's better to give credits to organisation supporting this work:
- `Funded by NAME`
- `Funded by URL`
- `Funded by NAME URL`
-->

Funded by BRGM & Ifremer.

fxprunayre committed:
Formatter / schema.org / Add croissant spec 🥐

"Croissant 🥐 is a high-level format for machine learning datasets that combines metadata, resource file descriptions, data structure, and default ML semantics into a single file; it works with existing datasets to make them easier to find, use, and support with tools. Croissant builds on schema.org, and its Dataset vocabulary, a widely used format to represent datasets on the Web, and make them searchable. "

Croissant is extending schema.org, this improvement review the current schema.org formatter to support additional 🥐 metadata available in ISO format. This is mainly about adding:
* croissant fileObject based on online resources with a download protocol
* croissant recordSet based on the feature catalogue

Refactor JSON-LD formatter for using same base formatter for both ISO19139 and ISO19115-3 to facilitate maintenance (similar to citation and DCAT formatter).

fxprunayre committed:
Formatter / JSON format / Ensure generated document is valid

Formatter producing JSON may produce invalid document as XSLT process output text which is written in the response. Ensure the JSON is valid and format it. Log any error, to be able to monitor them and improve the formatter for not well managed encoding.

In future version, consider using XSLT3 which support JSON output (https://www.w3.org/TR/xslt-30/#json).

fxprunayre committed:
Formatter / schema.org / Add croissant spec 🥐 / Add validator ref.

fxprunayre committed:
Formatter / schema.org / Add croissant spec 🥐 / Make the template more easy to add as dependency.

fxprunayre committed:
Formatter / JSONLD / Improvements

* Add basic dublin core support
* Improve JSON encoding

fxprunayre committed:
Formatter / JSONLD / Improvements

fxprunayre committed:
Formatter / JSONLD / test.

fxprunayre committed:
Formatter / JSONLD / Test updated for Dublin core.

jodygarnett reviewed (changes requested):
Please include docs in PR.

fxprunayre committed:
Enhance search engines section with robots.txt and sitemap info

Updated the search engines section with details on robots.txt and sitemap usage, including examples for better indexing.

jmckenna commented:
I might even expand that (as not everyone knows what "schema.org" is) to :"GeoNetwork includes a representation of that record in the schema.org (Structured Data for the Web) framework, encoded as [JSON-LD](https://en.wikipedia.org/wiki/JSON-LD), on any HTML representation of a *published* metadata record."

jmckenna commented:
And further expanded (as the JSON-LD is only embedded on API-rendered HTML pages, not "any" record's HTML page): "GeoNetwork includes a representation of that record in the schema.org (Structured Data for the Web) framework, encoded as [JSON-LD](https://en.wikipedia.org/wiki/JSON-LD), on any HTML representation (through the GeoNetwork API / sitemap) of a published metadata record."

fxprunayre committed:
Doc updates from review

josegar74 committed:
Merge branch 'main' into 44-croissant

josegar74 reviewed (commented):
There are 2 things unclear to me in this PR:

- It contains changes for bibtext and DCAT formatters, that I don't think are related to the PR.
- The Croissant spec changes are included in the JSON-LD formatter directly. Is this what we want?

fxprunayre commented:
> Out of interest, is this still on track @fxprunayre ?

I hope so.

@jodygarnett anything else you want to be improved here?

josegar74 committed:
Merge branch 'main' into 44-croissant

josegar74 commented on services/src/main/java/org/fao/geonet/api/records/formatters/PdfOrHtmlResponseWriter.java:
```diff
@@ -24,6 +26,22 @@ void writeOutResponse(ServiceContext context, String metadataUuid, String lang,
+                // XSLT is not the best way to generate JSON, so ensure the output is valid JSON.
+                // and log error to be able to improve the formatter if needed.
+                try {
+                    ObjectMapper mapper = new ObjectMapper();
```
ObjectMapper is thread-safe once configured, it can be reused across requests. a single shared instance.

fxprunayre replied:
Updated

fxprunayre committed:
Formatter / schema.org / Reuse ObjectMapper.

fxprunayre committed:
Merge remote-tracking branch 'origin/main' into 44-croissant

fxprunayre committed:
Formatter / schema.org / Add croissant spec / Test update.

fxprunayre committed:
Formatter / JSONLD use citation / Adapt test for https://github.com/geonetwork/core-geonetwork/pull/9191.

jodygarnett commented on services/src/main/java/org/fao/geonet/api/records/formatters/PdfOrHtmlResponseWriter.java:
```diff
@@ -15,6 +17,7 @@
 
 @Component
 class PdfOrHtmlResponseWriter {
```
```suggestion
/**
 * Responsible for transformations to `application/pdf`, `text/html`, and `text/json`.
 */
class PdfOrHtmlResponseWriter {
```

jodygarnett commented on docs/manual/docs/tutorials/introduction/extra/index.md:
````diff
@@ -20,6 +20,27 @@ Read more about the usage of Kibana at the [Kibana introduction](https://www.ela
+  </url>
+```
+
+GeoNetwork includes a representation of that record in [schema.org](https://schema.org/) (Structured Data for the Web) framework, encoded as [`json-ld`](https://en.wikipedia.org/wiki/JSON-LD), on any HTML representation (through the GeoNetwork API / sitemap) of a published metadata record. The `json-ld` representation is embedded in the html page as a script element and contains also [Croissant e ...
````
I am having some trouble understanding what "that record" refers to, I am assuimg the sitemap record above.

The paragraph also transitions from talking about the site map record ... to talking about any record? (please confirm).

```suggestion
GeoNetwork includes a representation of the sitemap record in [schema.org](https://schema.org/) (Structured Data for the Web) framework:

 *   Published metadata record accessed through sitemap / GeoNetwork API as HTML include [`json-ld`](https://en.wikipedia.org/wiki/JSON-LD) linked data (embedded JSON in the HTML page as a script element). 

 *   The included `json-ld` uses vocabulary from schema.org and also the [Croissant extension](https://docs.mlcommons.org/croissant/) for machine learning.

     > Croissant 🥐 is a high-level format for machine learning datasets that combines metadata, resource file descriptions, data structure, and default ML semantics into a single file; it works with existing datasets to make them easier to find, use, and support with tools. Croissant builds on schema.org, and its Dataset vocabulary, a widely used format to represent datasets on the Web, and make them searchable.
     
     Croissant is used to cover data model aspects (ISO19110) and additional details about distributions not covered by schema.org.
```

jodygarnett reviewed (changes requested):
@fxprunayre the documentation helped a lot thank you!

I did have some trouble understanding it as one big paragraph and have a suggestion on breaking it into bullet points (and using markdown formatting) to make it easier to understand.

My reading of the code is that schema.org and croissant extention applies to all metadata records, not only the sitemap (please confirm).

fxprunayre committed:
Apply suggestion from @jodygarnett

Co-authored-by: Jody Garnett <jody.garnett@gmail.com>

fxprunayre committed:
Apply suggestion from @jodygarnett

Co-authored-by: Jody Garnett <jody.garnett@gmail.com>

fxprunayre commented:
> My reading of the code is that schema.org and croissant extention applies to all metadata records, not only the sitemap (please confirm).

Yes. JSONLD formatter is embedded into the landing pages which are listed in the sitemap.

eg. 
https://datasetsearch.research.google.com/search?src=0&query=EEA%20reference%20grid%20for%20Europe&docid=L2cvMTFsa2dra2x3Mw%3D%3D
link to the landing page
https://sdi.eea.europa.eu/catalogue/geoss/api/records/0e97fc9e-9a3f-4e3e-80a2-92cc49fd4cc0f
which contains the JSONLD
https://sdi.eea.europa.eu/catalogue/geoss/api/records/0e97fc9e-9a3f-4e3e-80a2-92cc49fd4cc0f/formatters/jsonld
